Content.Client/Chat/ChatBox.cs

```diff
@@ -134,27 +508,227 @@ private void InputKeyBindDown(GUIBoundKeyEventArgs args)
+                }
+
+                ClampSize(left, bottom);
+            }
+        }
+
+        protected override void UIScaleChanged()
+        {
+            base.UIScaleChanged();
+            ClampAfterDelay();
+        }
+
+        private void ClydeOnOnWindowResized(WindowResizedEventArgs obj)
+        {
+            ClampAfterDelay();
+        }
+
+        private void ClampAfterDelay()
+        {
+            if (!_lobbyMode)
+                _clampIn = 2;
+        }
+
+        protected override void FrameUpdate(FrameEventArgs args)
+        {
+            base.FrameUpdate(args);
+            if (_lobbyMode) return;
+            // we do the clamping after a delay (after UI scale / window resize)
+            // because we need to wait for our parent container to properly resize
+            // first, so we can calculate where we should go. If we do it right away,
+            // we won't have the correct values from the parent to know how to adjust our margins.
+            if (_clampIn <= 0) return;
+            _clampIn -= 1;
+            if (_clampIn == 0) ClampSize();
+        }
+
+        private void ClampSize(float? desiredLeft = null, float? desiredBottom = null)
+        {
+            if (Parent == null || _lobbyMode) return;
+            var top = Rect.Top;
+            var right = Rect.Right;
+            var left = desiredLeft ?? Rect.Left;
+            var bottom = desiredBottom ?? Rect.Bottom;
+
+            // clamp so it doesn't go too high or low (leave space for alerts UI)
+            var maxBottom = Parent.Size.Y - MinDistanceFromBottom;
+            if (maxBottom <= MinHeight)
+            {
+                // we can't fit in our given space (window made awkwardly small), so give up
+                // and overlap at our min height
+                bottom = MinHeight;
+            }
+            else
+            {
+                bottom = Math.Clamp(bottom, MinHeight, maxBottom);
+            }
+
+            var maxLeft = Parent.Size.X - MinWidth;
+            if (maxLeft <= MinLeft)
+            {
+                // window too narrow, give up and overlap at our max left
+                left = maxLeft;
+            }
+            else
+            {
+                left = Math.Clamp(left, MinLeft, maxLeft);
+            }
+
+            LayoutContainer.SetMarginLeft(this, -((right + 10) - left));
+            LayoutContainer.SetMarginBottom(this, bottom);
+            OnResized?.Invoke(new ChatResizedEventArgs(bottom));
+        }
+
+        protected override void MouseExited()
+        {
+            if (_currentDrag == DragMode.None && !_lobbyMode)
+            {
+                DefaultCursorShape = CursorShape.Arrow;
+            }
         }
 
-        public void AddLine(FormattedMessage message, Color color)
+
+        private void InputOnTextChanged(LineEdit.LineEditEventArgs obj)
+        {
+            // switch temporarily to a different channel if an alias prefix has been entered.
```

Comment:
This function runs on all text entered even when the ChatBox is not focused. Needs a check for KeyboardFocused. Right now it's changing chat channel when any of the chat channel keys are pressed even when the chat is not focused.